### Importing required libraries

In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as tf
import cv2
import numpy as np
import os
import matplotlib.pyplot as plt

from tqdm.notebook import tqdm
import PIL
import cv2
%matplotlib inline

### Setting device as GPU

In [2]:
if torch.cuda.is_available():
    print("GPU available:", torch.cuda.get_device_name())
    device = torch.device('cuda:0')
else:
    print("******Workking on CPU*******")
    device = torch.device('cpu')
device

GPU available: GeForce GTX 1070


device(type='cuda', index=0)

###  Prepare data

In [3]:
CROP_SIZE = 128
class TrainDataset(Dataset):
    def __init__(self, path):
        super(TrainDataset, self).__init__()
        self.image_names = [os.path.join(path,x) for x in os.listdir(path)]
        self.lr_preprocess = tf.Compose([tf.ToPILImage(),tf.Resize((CROP_SIZE,CROP_SIZE), interpolation = PIL.Image.BICUBIC),tf.ToTensor()])
        self.hr_preprocess = tf.Compose([tf.ToTensor()])
        self.len = len(self.image_names)
    def __len__(self):
        return self.len
    def __getitem__(self, index):
        img = self.image_names[index]
        hr = self.hr_preprocess(PIL.Image.open(img))
        lr = self.lr_preprocess(hr)
        return lr,hr

lr_preprocess = tf.Compose([tf.ToPILImage(),tf.Resize((CROP_SIZE,CROP_SIZE), interpolation = PIL.Image.BICUBIC),tf.ToTensor()])
hr_preprocess = tf.Compose([tf.ToTensor()])
img =PIL.Image.open(path)
img_hr = hr_preprocess(PIL.Image.open(path))
img_lr = lr_preprocess(img_hr)

# Model

<img src = "https://miro.medium.com/max/4916/1*zsiBj3IL4ALeLgsCeQ3lyA.png">

### Generator Network

In [16]:
#Residual Block
class Residual_Block(nn.Module):
    def __init__(self):
        super(Residual_Block, self).__init__()
        self.net = nn.Sequential(
        nn.Conv2d(in_channels = 64, out_channels = 64, kernel_size = 3, padding = 1),
        nn.BatchNorm2d(num_features = 64),
        nn.PReLU(),
        nn.Conv2d(in_channels = 64, out_channels = 64, kernel_size = 3, padding = 1),
        nn.BatchNorm2d(num_features = 64)
        )
    def forward(self, x):
        x = x+self.net(x)
        #print("After residial block:",x.shape)
        return x
#Upsample Block
class Upsample_Block(nn.Module):
    def __init__(self, in_ch,scale_factor):
        super(Upsample_Block, self).__init__()
        self.net  = nn.Sequential(
        nn.Conv2d(in_channels = in_ch, out_channels = in_ch*(scale_factor**2), kernel_size = 3, stride = 1, padding = 1),
        nn.PixelShuffle(upscale_factor = scale_factor),
        nn.PReLU()
        )
    def forward(self, x):
        x = self.net(x)
        #print("After upsampling:",x.shape)
        return x
        

#Generator Block
class Generator(nn.Module):
    def __init__(self):
        super(Generator, self).__init__()
        #First Block
        self.block1 = nn.Sequential(
        nn.Conv2d(in_channels = 3, out_channels = 64, kernel_size = 9, padding = 4,stride = 1),
        nn.PReLU()
        )
        #8 Residual Blocks
        self.residuals = nn.ModuleList([Residual_Block() for x in range(8)]) 
        self.block2 = nn.Sequential(
        nn.Conv2d(in_channels = 64, out_channels = 64, kernel_size = 3, stride = 1,padding = 1),
        nn.BatchNorm2d(num_features = 64)
        )
        # 2 Upsampling Blocks
        self.upsamples = nn.Sequential(
        Upsample_Block(64,2),
        Upsample_Block(64,2),
        nn.Conv2d(in_channels = 64, out_channels = 3, kernel_size = 9, stride = 1, padding = 4)
        )
    def forward(self, x):
        #print("Initial shape:",x.shape)
        #print("Device:",device)
        x = self.block1(x)
        #print("After Block 1:", x.shape)
        x_res = self.residuals[0](x)
        for i in range(7):
            x_res = self.residuals[i+1](x_res)
        x_res = self.block2(x)
        x = x+x_res
        x = self.upsamples(x)
        #print("Output Shape:",x.shape)
        return (torch.tanh(x)+1)/2

### Discriminator Network

In [5]:
class Discriminator(nn.Module):
    def __init__(self, l=0.2):
        super(Discriminator, self).__init__()
        self.net = nn.Sequential(
        nn.Conv2d(in_channels = 3, out_channels = 64, kernel_size = 3, stride = 1),
        nn.LeakyReLU(l),
        
        nn.Conv2d(in_channels = 64, out_channels = 64, kernel_size = 3, stride = 2, padding = 1),
        nn.BatchNorm2d(num_features = 64),
        nn.LeakyReLU(l),
            
        nn.Conv2d(in_channels = 64, out_channels = 128, kernel_size = 3, stride = 1, padding = 1),
        nn.BatchNorm2d(num_features = 128),
        nn.LeakyReLU(l),
        
        nn.Conv2d(in_channels = 128, out_channels = 128, kernel_size = 3, stride = 2, padding = 1),
        nn.BatchNorm2d(num_features = 128),
        nn.LeakyReLU(l),
            
        nn.Conv2d(in_channels = 128, out_channels = 256, kernel_size = 3, stride = 1, padding = 1),
        nn.BatchNorm2d(num_features = 256),
        nn.LeakyReLU(l),
            
        nn.Conv2d(in_channels = 256, out_channels = 256, kernel_size = 3, stride = 2, padding = 1),
        nn.BatchNorm2d(num_features = 256),
        nn.LeakyReLU(l),
            
        nn.Conv2d(in_channels = 256, out_channels = 512, kernel_size = 3, stride = 1, padding = 1),
        nn.BatchNorm2d(num_features = 512),
        nn.LeakyReLU(l),
            
        nn.Conv2d(in_channels = 512, out_channels = 512, kernel_size = 3, stride = 2, padding = 1),
        nn.BatchNorm2d(num_features = 512),
        nn.LeakyReLU(l),
            
        nn.AdaptiveAvgPool2d(1),
        nn.Conv2d(512, 1024, kernel_size = 1),
        nn.LeakyReLU(l),
        nn.Conv2d(1024,1,kernel_size = 1)
        )
    def forward(self,x):
        y = self.net(x)
        yhat = torch.sigmoid(y).view(x.shape[0])
        return yhat

Lets test the dimensional correctness with a random vector.

In [6]:
x = torch.randn((5,3,512,512))
disc = Discriminator()
y = disc(x)
y.data.shape

torch.Size([5])

Look Good!

##  Training Procedure

The generic method of training GANs is given bellow. It was proposed by Ian Goodfellow et al in the first paper that introduced GANs.

<img src = "https://3qeqpr26caki16dnhd19sv6by6v-wpengine.netdna-ssl.com/wp-content/uploads/2019/05/Summary-of-the-Generative-Adversarial-Network-Training-Algorithm-1024x669.png" width = 600>

### Pretraining Generator

In [40]:
def pre_train(loader):
    optimizerG = torch.optim.Adam(netG.parameters())
    print("Pretraining Generator:")
    for epoch in range(PRETRAIN_EPOCHS):
        print("\tEpoch: ",epoch)
        netG.train()
        for lr, hr in tqdm(loader):
            lr = lr.cuda()
            hr = hr.cuda()
            sr  = netG(lr)
            netG.zero_grad()
            loss = mse(sr,hr)
            loss.backward()
            optimizerG.step()

### Training GAN

In [71]:
def train_GAN(loader):
    print("Training GAN:")
    for epoch in range(EPOCHS):
        print("\tEpoch: ", epoch)
        netG.train()
        netD.train()
        for lr,hr in tqdm(loader):
            lr = lr.cuda()
            hr = hr.cuda()
            #Train Discriminator
            netD.zero_grad()
            
            logits_real = netD(hr)
            logits_fake = netD(netG(lr).detach())
            
            logits_real = logits_real.cuda()
            logits_fake = logits_fake.cuda()
            #Label Smoothing
            real = torch.randn(logits_real.size())*0.25+0.85
            fake = torch.randn(logits_fake.size())*0.15
            
            real = real.cuda()
            fake = fake.cuda()
            #Label Flipping (not implemented...yet)
            d_loss = bce(logits_real,real) + bce(logits_fake, fake)
            d_loss.backward()
            optimizerD.step()
            
            #Train Generator
            netG.zero_grad()
            sr = netG(lr)
            content_loss = mse(sr,hr)
            
            logits_fake_new = netD(sr)
            adverserial_loss = bce(logits_fake_new, torch.ones_like(logits_fake_new))
            g_loss = content_loss + 1e-2*adverserial_loss
            g_loss.backward()
            optimizerG.step()
                               

In [ ]:
#path = os.path.join(os.getcwd(),"data")
#path = "D:\\Games Setup\\data"


Checking if Dataset object works as expected-

In [ ]:
# for lr,hr in train_set:
#     hr = np.transpose(hr.numpy(),(1,2,0))[:,:,::-1]
#     lr = np.transpose(lr.numpy(),(1,2,0))[:,:,::-1]
#     #print("HR shape:",hr.shape)
#     cv2.imshow("HR Image",hr)
#     cv2.waitKey(0)
#     cv2.destroyAllWindows()
#     cv2.imshow("LR Image",lr)
#     cv2.waitKey(0)
#     cv2.destroyAllWindows()

It does, great!

In [85]:
HR_SIZE = 512
LR_SIZE = 128
SCALING_FACTOR = 4
PRETRAIN_EPOCHS = 2
EPOCHS = 1000

I will first write the basic backbone training function, then add other functionalities such as-
<ul>
    <li>Sending model to GPU
    <li>Label Smoothing(well, I did it anyways)</li> 
    <li>Label Flipping</li>
    <li>Saving checkpoints when training</li>
    <li>Visualizing the model with Tensorboard</li>
    <li>Displaying epochwise performance</li>

In [72]:
netG = Generator()
netD = Discriminator()
optimizerD = torch.optim.Adam(netD.parameters())
optimizerG = torch.optim.Adam(netG.parameters())
mse = nn.MSELoss()
bce = nn.BCELoss()
pretrain = True
train_path = "D:\\Games Setup\\FFHQ\\train"
train_set = TrainDataset(train_path)
test_path =  "D:\\Games Setup\\FFHQ\\test"
test_set = TrainDataset(test_path)
train_loader = DataLoader(train_set, batch_size = 10)
test_loader = DataLoader(test_set, batch_size = 100)

In [73]:
netG.cuda()
netD.cuda()
mse.cuda()
bce.cuda()

BCELoss()

### Passing debug data through our functions first.

In [74]:
import timeit
import time

#### Timing pretrainer

In [77]:
debug_path = "D:\\Games Setup\\data"
debug_set = TrainDataset(debug_path)
debug_loader = DataLoader(debug_set)

In [84]:
start_time = time.time()
pre_train(debug_loader)
end_time = time.time()
print("Time to pretrain 1 image for ",PRETRAIN_EPOCHS, "= ",(end_time-start_time)*100,"ms")

Pretraining Generator:
	Epoch:  0



	Epoch:  1



Time to pretrain 1 image for  2 =  52.59981155395508 ms


####  Timing GAN trainer

In [88]:
start_time = time.time()
train_GAN(debug_loader)
end_time = time.time()
print("Time to train GAN on 1 image for ",EPOCHS, " epochs = ",(end_time-start_time),"s")

Training GAN:
	Epoch:  0



	Epoch:  1



	Epoch:  2



	Epoch:  3



	Epoch:  4



	Epoch:  5



	Epoch:  6



	Epoch:  7



	Epoch:  8



	Epoch:  9



	Epoch:  10



	Epoch:  11



	Epoch:  12



	Epoch:  13



	Epoch:  14



	Epoch:  15



	Epoch:  16



	Epoch:  17



	Epoch:  18



	Epoch:  19



	Epoch:  20



	Epoch:  21



	Epoch:  22



	Epoch:  23



	Epoch:  24



	Epoch:  25



	Epoch:  26



	Epoch:  27



	Epoch:  28



	Epoch:  29



	Epoch:  30



	Epoch:  31



	Epoch:  32



	Epoch:  33



	Epoch:  34



	Epoch:  35



	Epoch:  36



	Epoch:  37



	Epoch:  38



	Epoch:  39



	Epoch:  40



	Epoch:  41



	Epoch:  42



	Epoch:  43



	Epoch:  44



	Epoch:  45



	Epoch:  46



	Epoch:  47



	Epoch:  48



	Epoch:  49



	Epoch:  50



	Epoch:  51



	Epoch:  52



	Epoch:  53



	Epoch:  54



	Epoch:  55



	Epoch:  56



	Epoch:  57



	Epoch:  58



	Epoch:  59



	Epoch:  60



	Epoch:  61



	Epoch:  62



	Epoch:  63



	Epoch:  64



	Epoch:  65



	Epoch:  66



	Epoch:  67



	Epoch:  68



	Epoch:  69



	Epoch:  70



	Epoch:  71



	Epoch:  72



	Epoch:  73



	Epoch:  74



	Epoch:  75



	Epoch:  76



	Epoch:  77



	Epoch:  78



	Epoch:  79



	Epoch:  80



	Epoch:  81



	Epoch:  82



	Epoch:  83



	Epoch:  84



	Epoch:  85



	Epoch:  86



	Epoch:  87



	Epoch:  88



	Epoch:  89



	Epoch:  90



	Epoch:  91



	Epoch:  92



	Epoch:  93



	Epoch:  94



	Epoch:  95



	Epoch:  96



	Epoch:  97



	Epoch:  98



	Epoch:  99



	Epoch:  100



	Epoch:  101



	Epoch:  102



	Epoch:  103



	Epoch:  104



	Epoch:  105



	Epoch:  106



	Epoch:  107



	Epoch:  108



	Epoch:  109



	Epoch:  110



	Epoch:  111



	Epoch:  112



	Epoch:  113



	Epoch:  114



	Epoch:  115



	Epoch:  116



	Epoch:  117



	Epoch:  118



	Epoch:  119



	Epoch:  120



	Epoch:  121



	Epoch:  122



	Epoch:  123



	Epoch:  124



	Epoch:  125



	Epoch:  126



	Epoch:  127



	Epoch:  128



	Epoch:  129



	Epoch:  130



	Epoch:  131



	Epoch:  132



	Epoch:  133



	Epoch:  134



	Epoch:  135



	Epoch:  136



	Epoch:  137



	Epoch:  138



	Epoch:  139



	Epoch:  140



	Epoch:  141



	Epoch:  142



	Epoch:  143



	Epoch:  144



	Epoch:  145



	Epoch:  146



	Epoch:  147



	Epoch:  148



	Epoch:  149



	Epoch:  150



	Epoch:  151



	Epoch:  152



	Epoch:  153



	Epoch:  154



	Epoch:  155



	Epoch:  156



	Epoch:  157



	Epoch:  158



	Epoch:  159



	Epoch:  160



	Epoch:  161



	Epoch:  162



	Epoch:  163



	Epoch:  164



	Epoch:  165



	Epoch:  166



	Epoch:  167



	Epoch:  168



	Epoch:  169



	Epoch:  170



	Epoch:  171



	Epoch:  172



	Epoch:  173



	Epoch:  174



	Epoch:  175



	Epoch:  176



	Epoch:  177



	Epoch:  178



	Epoch:  179



	Epoch:  180



	Epoch:  181



	Epoch:  182



	Epoch:  183



	Epoch:  184



	Epoch:  185



	Epoch:  186



	Epoch:  187



	Epoch:  188



	Epoch:  189



	Epoch:  190



	Epoch:  191



	Epoch:  192



	Epoch:  193



	Epoch:  194



	Epoch:  195



	Epoch:  196



	Epoch:  197



	Epoch:  198



	Epoch:  199



	Epoch:  200



	Epoch:  201



	Epoch:  202



	Epoch:  203



	Epoch:  204



	Epoch:  205



	Epoch:  206



	Epoch:  207



	Epoch:  208



	Epoch:  209



	Epoch:  210



	Epoch:  211



	Epoch:  212



	Epoch:  213



	Epoch:  214



	Epoch:  215



	Epoch:  216



	Epoch:  217



	Epoch:  218



	Epoch:  219



	Epoch:  220



	Epoch:  221



	Epoch:  222



	Epoch:  223



	Epoch:  224



	Epoch:  225



	Epoch:  226



	Epoch:  227



	Epoch:  228



	Epoch:  229



	Epoch:  230



	Epoch:  231



	Epoch:  232



	Epoch:  233



	Epoch:  234



	Epoch:  235



	Epoch:  236



	Epoch:  237



	Epoch:  238



	Epoch:  239



	Epoch:  240



	Epoch:  241



	Epoch:  242



	Epoch:  243



	Epoch:  244



	Epoch:  245



	Epoch:  246



	Epoch:  247



	Epoch:  248



	Epoch:  249



	Epoch:  250



	Epoch:  251



	Epoch:  252



	Epoch:  253



	Epoch:  254



	Epoch:  255



	Epoch:  256



	Epoch:  257



	Epoch:  258



	Epoch:  259



	Epoch:  260



	Epoch:  261



	Epoch:  262



	Epoch:  263



	Epoch:  264



	Epoch:  265



	Epoch:  266



	Epoch:  267



	Epoch:  268



	Epoch:  269



	Epoch:  270



	Epoch:  271



	Epoch:  272



	Epoch:  273



	Epoch:  274



	Epoch:  275



	Epoch:  276



	Epoch:  277



	Epoch:  278



	Epoch:  279



	Epoch:  280



	Epoch:  281



	Epoch:  282



	Epoch:  283



	Epoch:  284



	Epoch:  285



	Epoch:  286



	Epoch:  287



	Epoch:  288



	Epoch:  289



	Epoch:  290



	Epoch:  291



	Epoch:  292



	Epoch:  293



	Epoch:  294



	Epoch:  295



	Epoch:  296



	Epoch:  297



	Epoch:  298



	Epoch:  299



	Epoch:  300



	Epoch:  301



	Epoch:  302



	Epoch:  303



	Epoch:  304



	Epoch:  305



	Epoch:  306



	Epoch:  307



	Epoch:  308



	Epoch:  309



	Epoch:  310



	Epoch:  311



	Epoch:  312



	Epoch:  313



	Epoch:  314



	Epoch:  315



	Epoch:  316



	Epoch:  317



	Epoch:  318



	Epoch:  319



	Epoch:  320



	Epoch:  321



	Epoch:  322



	Epoch:  323



	Epoch:  324



	Epoch:  325



	Epoch:  326



	Epoch:  327



	Epoch:  328



	Epoch:  329



	Epoch:  330



	Epoch:  331



	Epoch:  332



	Epoch:  333



	Epoch:  334



	Epoch:  335



	Epoch:  336



	Epoch:  337



	Epoch:  338



	Epoch:  339



	Epoch:  340



	Epoch:  341



	Epoch:  342



	Epoch:  343



	Epoch:  344



	Epoch:  345



	Epoch:  346



	Epoch:  347



	Epoch:  348



	Epoch:  349



	Epoch:  350



	Epoch:  351



	Epoch:  352



	Epoch:  353



	Epoch:  354



	Epoch:  355



	Epoch:  356



	Epoch:  357



	Epoch:  358



	Epoch:  359



	Epoch:  360



	Epoch:  361



	Epoch:  362



	Epoch:  363



	Epoch:  364



	Epoch:  365



	Epoch:  366



	Epoch:  367



	Epoch:  368



	Epoch:  369



	Epoch:  370



	Epoch:  371



	Epoch:  372



	Epoch:  373



	Epoch:  374



	Epoch:  375



	Epoch:  376



	Epoch:  377



	Epoch:  378



	Epoch:  379



	Epoch:  380



	Epoch:  381



	Epoch:  382



	Epoch:  383



	Epoch:  384



	Epoch:  385



	Epoch:  386



	Epoch:  387



	Epoch:  388



	Epoch:  389



	Epoch:  390



	Epoch:  391



	Epoch:  392



	Epoch:  393



	Epoch:  394



	Epoch:  395



	Epoch:  396



	Epoch:  397



	Epoch:  398



	Epoch:  399



	Epoch:  400



	Epoch:  401



	Epoch:  402



	Epoch:  403



	Epoch:  404



	Epoch:  405



	Epoch:  406



	Epoch:  407



	Epoch:  408



	Epoch:  409



	Epoch:  410



	Epoch:  411



	Epoch:  412



	Epoch:  413



	Epoch:  414



	Epoch:  415



	Epoch:  416



	Epoch:  417



	Epoch:  418



	Epoch:  419



	Epoch:  420



	Epoch:  421



	Epoch:  422



	Epoch:  423



	Epoch:  424



	Epoch:  425



	Epoch:  426



	Epoch:  427



	Epoch:  428



	Epoch:  429



	Epoch:  430



	Epoch:  431



	Epoch:  432



	Epoch:  433



	Epoch:  434



	Epoch:  435



	Epoch:  436



	Epoch:  437



	Epoch:  438



	Epoch:  439



	Epoch:  440



	Epoch:  441



	Epoch:  442



	Epoch:  443



	Epoch:  444



	Epoch:  445



	Epoch:  446



	Epoch:  447



	Epoch:  448



	Epoch:  449



	Epoch:  450



	Epoch:  451



	Epoch:  452



	Epoch:  453



	Epoch:  454



	Epoch:  455



	Epoch:  456



	Epoch:  457



	Epoch:  458



	Epoch:  459



	Epoch:  460



	Epoch:  461



	Epoch:  462



	Epoch:  463



	Epoch:  464



	Epoch:  465



	Epoch:  466



	Epoch:  467



	Epoch:  468



	Epoch:  469



	Epoch:  470



	Epoch:  471



	Epoch:  472



	Epoch:  473



	Epoch:  474



	Epoch:  475



	Epoch:  476



	Epoch:  477



	Epoch:  478



	Epoch:  479



	Epoch:  480



	Epoch:  481



	Epoch:  482



	Epoch:  483



	Epoch:  484



	Epoch:  485



	Epoch:  486



	Epoch:  487



	Epoch:  488



	Epoch:  489



	Epoch:  490



	Epoch:  491



	Epoch:  492



	Epoch:  493



	Epoch:  494



	Epoch:  495



	Epoch:  496



	Epoch:  497



	Epoch:  498



	Epoch:  499



	Epoch:  500



	Epoch:  501



	Epoch:  502



	Epoch:  503



	Epoch:  504



	Epoch:  505



	Epoch:  506



	Epoch:  507



	Epoch:  508



	Epoch:  509



	Epoch:  510



	Epoch:  511



	Epoch:  512



	Epoch:  513



	Epoch:  514



	Epoch:  515



	Epoch:  516



	Epoch:  517



	Epoch:  518



	Epoch:  519



	Epoch:  520



	Epoch:  521



	Epoch:  522



	Epoch:  523



	Epoch:  524



	Epoch:  525



	Epoch:  526



	Epoch:  527



	Epoch:  528



	Epoch:  529



	Epoch:  530



	Epoch:  531



	Epoch:  532



	Epoch:  533



	Epoch:  534



	Epoch:  535



	Epoch:  536



	Epoch:  537



	Epoch:  538



	Epoch:  539



	Epoch:  540



	Epoch:  541



	Epoch:  542



	Epoch:  543



	Epoch:  544



	Epoch:  545



	Epoch:  546



	Epoch:  547



	Epoch:  548



	Epoch:  549



	Epoch:  550



	Epoch:  551



	Epoch:  552



	Epoch:  553



	Epoch:  554



	Epoch:  555



	Epoch:  556



	Epoch:  557



	Epoch:  558



	Epoch:  559



	Epoch:  560



	Epoch:  561



	Epoch:  562



	Epoch:  563



	Epoch:  564



	Epoch:  565



	Epoch:  566



	Epoch:  567



	Epoch:  568



	Epoch:  569



	Epoch:  570



	Epoch:  571



	Epoch:  572



	Epoch:  573



	Epoch:  574



	Epoch:  575



	Epoch:  576



	Epoch:  577



	Epoch:  578



	Epoch:  579



	Epoch:  580



	Epoch:  581



	Epoch:  582



	Epoch:  583



	Epoch:  584



	Epoch:  585



	Epoch:  586



	Epoch:  587



	Epoch:  588



	Epoch:  589



	Epoch:  590



	Epoch:  591



	Epoch:  592



	Epoch:  593



	Epoch:  594



	Epoch:  595



	Epoch:  596



	Epoch:  597



	Epoch:  598



	Epoch:  599



	Epoch:  600



	Epoch:  601



	Epoch:  602



	Epoch:  603



	Epoch:  604



	Epoch:  605



	Epoch:  606



	Epoch:  607



	Epoch:  608



	Epoch:  609



	Epoch:  610



	Epoch:  611



	Epoch:  612



	Epoch:  613



	Epoch:  614



	Epoch:  615



	Epoch:  616



	Epoch:  617



	Epoch:  618



	Epoch:  619



	Epoch:  620



	Epoch:  621



	Epoch:  622



	Epoch:  623



	Epoch:  624



	Epoch:  625



	Epoch:  626



	Epoch:  627



	Epoch:  628



	Epoch:  629



	Epoch:  630



	Epoch:  631



	Epoch:  632



	Epoch:  633



	Epoch:  634



	Epoch:  635



	Epoch:  636



	Epoch:  637



	Epoch:  638



	Epoch:  639



	Epoch:  640



	Epoch:  641



	Epoch:  642



	Epoch:  643



	Epoch:  644



	Epoch:  645



	Epoch:  646



	Epoch:  647



	Epoch:  648



	Epoch:  649



	Epoch:  650



	Epoch:  651



	Epoch:  652



	Epoch:  653



	Epoch:  654



	Epoch:  655



	Epoch:  656



	Epoch:  657



	Epoch:  658



	Epoch:  659



	Epoch:  660



	Epoch:  661



	Epoch:  662



	Epoch:  663



	Epoch:  664



	Epoch:  665



	Epoch:  666



	Epoch:  667



	Epoch:  668



	Epoch:  669



	Epoch:  670



	Epoch:  671



	Epoch:  672



	Epoch:  673



	Epoch:  674



	Epoch:  675



	Epoch:  676



	Epoch:  677



	Epoch:  678



	Epoch:  679



	Epoch:  680



	Epoch:  681



	Epoch:  682



	Epoch:  683



	Epoch:  684



	Epoch:  685



	Epoch:  686



	Epoch:  687



	Epoch:  688



	Epoch:  689



	Epoch:  690



	Epoch:  691



	Epoch:  692



	Epoch:  693



	Epoch:  694



	Epoch:  695



	Epoch:  696



	Epoch:  697



	Epoch:  698



	Epoch:  699



	Epoch:  700



	Epoch:  701



	Epoch:  702



	Epoch:  703



	Epoch:  704



	Epoch:  705



	Epoch:  706



	Epoch:  707



	Epoch:  708



	Epoch:  709



	Epoch:  710



	Epoch:  711



	Epoch:  712



	Epoch:  713



	Epoch:  714



	Epoch:  715



	Epoch:  716



	Epoch:  717



	Epoch:  718



	Epoch:  719



	Epoch:  720



	Epoch:  721



	Epoch:  722



	Epoch:  723



	Epoch:  724



	Epoch:  725



	Epoch:  726



	Epoch:  727



	Epoch:  728



	Epoch:  729



	Epoch:  730



	Epoch:  731



	Epoch:  732



	Epoch:  733



	Epoch:  734



	Epoch:  735



	Epoch:  736



	Epoch:  737



	Epoch:  738



	Epoch:  739



	Epoch:  740



	Epoch:  741



	Epoch:  742



	Epoch:  743



	Epoch:  744



	Epoch:  745



	Epoch:  746



	Epoch:  747



	Epoch:  748



	Epoch:  749



	Epoch:  750



	Epoch:  751



	Epoch:  752



	Epoch:  753



	Epoch:  754



	Epoch:  755



	Epoch:  756



	Epoch:  757



	Epoch:  758



	Epoch:  759



	Epoch:  760



	Epoch:  761



	Epoch:  762



	Epoch:  763



	Epoch:  764



	Epoch:  765



	Epoch:  766



	Epoch:  767



	Epoch:  768



	Epoch:  769



	Epoch:  770



	Epoch:  771



	Epoch:  772



	Epoch:  773



	Epoch:  774



	Epoch:  775



	Epoch:  776



	Epoch:  777



	Epoch:  778



	Epoch:  779



	Epoch:  780



	Epoch:  781



	Epoch:  782



	Epoch:  783



	Epoch:  784



	Epoch:  785



	Epoch:  786



	Epoch:  787



	Epoch:  788



	Epoch:  789



	Epoch:  790



	Epoch:  791



	Epoch:  792



	Epoch:  793



	Epoch:  794



	Epoch:  795



	Epoch:  796



	Epoch:  797



	Epoch:  798



	Epoch:  799



	Epoch:  800



	Epoch:  801



	Epoch:  802



	Epoch:  803



	Epoch:  804



	Epoch:  805



	Epoch:  806



	Epoch:  807



	Epoch:  808



	Epoch:  809



	Epoch:  810



	Epoch:  811



	Epoch:  812



	Epoch:  813



	Epoch:  814



	Epoch:  815



	Epoch:  816



	Epoch:  817



	Epoch:  818



	Epoch:  819



	Epoch:  820



	Epoch:  821



	Epoch:  822



	Epoch:  823



	Epoch:  824



	Epoch:  825



	Epoch:  826



	Epoch:  827



	Epoch:  828



	Epoch:  829



	Epoch:  830



	Epoch:  831



	Epoch:  832



	Epoch:  833



	Epoch:  834



	Epoch:  835



	Epoch:  836



	Epoch:  837



	Epoch:  838



	Epoch:  839



	Epoch:  840



	Epoch:  841



	Epoch:  842



	Epoch:  843



	Epoch:  844



	Epoch:  845



	Epoch:  846



	Epoch:  847



	Epoch:  848



	Epoch:  849



	Epoch:  850



	Epoch:  851



	Epoch:  852



	Epoch:  853



	Epoch:  854



	Epoch:  855



	Epoch:  856



	Epoch:  857



	Epoch:  858



	Epoch:  859



	Epoch:  860



	Epoch:  861



	Epoch:  862



	Epoch:  863



	Epoch:  864



	Epoch:  865



	Epoch:  866



	Epoch:  867



	Epoch:  868



	Epoch:  869



	Epoch:  870



	Epoch:  871



	Epoch:  872



	Epoch:  873



	Epoch:  874



	Epoch:  875



	Epoch:  876



	Epoch:  877



	Epoch:  878



	Epoch:  879



	Epoch:  880



	Epoch:  881



	Epoch:  882



	Epoch:  883



	Epoch:  884



	Epoch:  885



	Epoch:  886



	Epoch:  887



	Epoch:  888



	Epoch:  889



	Epoch:  890



	Epoch:  891



	Epoch:  892



	Epoch:  893



	Epoch:  894



	Epoch:  895



	Epoch:  896



	Epoch:  897



	Epoch:  898



	Epoch:  899



	Epoch:  900



	Epoch:  901



	Epoch:  902



	Epoch:  903



	Epoch:  904



	Epoch:  905



	Epoch:  906



	Epoch:  907



	Epoch:  908



	Epoch:  909



	Epoch:  910



	Epoch:  911



	Epoch:  912



	Epoch:  913



	Epoch:  914



	Epoch:  915



	Epoch:  916



	Epoch:  917



	Epoch:  918



	Epoch:  919



	Epoch:  920



	Epoch:  921



	Epoch:  922



	Epoch:  923



	Epoch:  924



	Epoch:  925



	Epoch:  926



	Epoch:  927



	Epoch:  928



	Epoch:  929



	Epoch:  930



	Epoch:  931



	Epoch:  932



	Epoch:  933



	Epoch:  934



	Epoch:  935



	Epoch:  936



	Epoch:  937



	Epoch:  938



	Epoch:  939



	Epoch:  940



	Epoch:  941



	Epoch:  942



	Epoch:  943



	Epoch:  944



	Epoch:  945



	Epoch:  946



	Epoch:  947



	Epoch:  948



	Epoch:  949



	Epoch:  950



	Epoch:  951



	Epoch:  952



	Epoch:  953



	Epoch:  954



	Epoch:  955



	Epoch:  956



	Epoch:  957



	Epoch:  958



	Epoch:  959



	Epoch:  960



	Epoch:  961



	Epoch:  962



	Epoch:  963



	Epoch:  964



	Epoch:  965



	Epoch:  966



	Epoch:  967



	Epoch:  968



	Epoch:  969



	Epoch:  970



	Epoch:  971



	Epoch:  972



	Epoch:  973



	Epoch:  974



	Epoch:  975



	Epoch:  976



	Epoch:  977



	Epoch:  978



	Epoch:  979



	Epoch:  980



	Epoch:  981



	Epoch:  982



	Epoch:  983



	Epoch:  984



	Epoch:  985



	Epoch:  986



	Epoch:  987



	Epoch:  988



	Epoch:  989



	Epoch:  990



	Epoch:  991



	Epoch:  992



	Epoch:  993



	Epoch:  994



	Epoch:  995



	Epoch:  996



	Epoch:  997



	Epoch:  998



	Epoch:  999



Time to train GAN on 1 image for  1000  epochs =  446.95525336265564 s


Time to train GAN on 1 image for  1000  epochs =  446.95525336265564 s

Holy fuck! Training 1 image takes 7 minutes. Training 50,000 images would take 250 days. Bloody Hell!